### DISCUSSION TITLE
効率のよい検索を利用する

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#57] [General]
## 議題
サブクエリを引数に取る場合、IN よりも EXISTS を使う

## なぜ

```sql
①SELECT * FROM parent WHERE id IN (SELECT id FROM children)
②SELECT * FROM parent WHERE id EXISTS (SELECT * FROM children WHERE parent.id = children.id)
```

① のほうが早く動作するから

- もし結合キーにインデックスが貼られていれば、children テーブルを見に行かず、index を見に行くだけでいい
- EXISTS は一行でも条件に合致する行を見つけたらそこで打ち切るので、IN のように全表検索の必要がない

## 実際に調べて見た結果

```sql
①EXPLAIN SELECT * FROM parent WHERE id in (select id from children);
②EXPLAIN SELECT * FROM parent WHERE exists (select * from children where parent.id = children.id);
③EXPLAIN SELECT parent.id, parent.name FROM parent INNER JOIN children ON parent.id = children.id;
```

どの実行計画も同じだった。
| id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | SIMPLE | parent | NULL | ALL | PRIMARY | NULL | NULL | NULL | 2 | 100.00 | NULL |
| 1 | SIMPLE | children | NULL | eq_ref | PRIMARY | PRIMARY | 4 | mysql_sandbox.parent.id | 1 | 100.00 | Using index |

実行順序も同じだった
①parentを全件取得
②childrenの情報をindex経由で取得
③parentとchildrenをNested loop結合
```
-> Nested loop inner join  (cost=1.6 rows=3) (actual time=0.0578..0.0734 rows=2 loops=1)
    -> Table scan on parent  (cost=0.55 rows=3) (actual time=0.0395..0.0461 rows=3 loops=1)
    -> Single-row covering index lookup on children using PRIMARY (id=parent.id)  (cost=0.283 rows=1) (actual time=0.0076..0.00767 rows=0.667 loops=3)
```

このことから

- もし結合キーにインデックスが貼られていれば、children テーブルを見に行かず、index を見に行くだけでいい
  - => in, existsに関わらず、インデックスが貼られていればchildren テーブルは見にいかない。Extra=Using index がそれを示している
- EXISTS は一行でも条件に合致する行を見つけたらそこで打ち切るので、IN のように全表検索の必要がない
  - => 今回の例では検索対象に PRIMARY KEY (PRIMARY KEYにはuniqなindexのような役割がある)を使っており、uniq な値のため 1 検索 1 行ということが明確になり、IN でも全表検索されない。type=eq_ref がそれを示している。-[参考文献](https://dev.mysql.com/doc/refman/8.0/ja/explain-output.html#explain-join-types)
- 実行計画が統一され、どの方法を選んでも内部的には変わらなくなった
  - => 古いバージョンだとINの中でサブクエリを実行すると、indexが効かなかったらしい-[参考文献](https://qiita.com/SSKNOK/items/99ed6a8734ac547357f7)

## 結論

- 今回の例では IN と EXISTS に速度の差はない。それならば直感的に理解しやすい、INを使うほうがよい
- 検索対象に PRIMARY KEYを使った時は、全表検索されないため、INとEXISTSを使い分ける必要がない。typeは以下の順で早くなる
  - eq_ref
  - index
  - all

## 補足

- 統計情報によって、同じ SQL でも実行計画は異なる場合がある
  - https://tech.spacely.co.jp/entry/2022/06/07/131350
  - https://qiita.com/NagaokaKenichi/items/5b6eb9887f88046a594d

**Comment by yuyakawai0411** [comment]
## 議題
Indexを見に行くとなぜ早くなるのか？

```sql
SELECT * FROM parent WHERE name = "Jhon";
```
## indexを貼っていない場合
**実行計画**
| id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | SIMPLE | parent | NULL | ALL | NULL | NULL | NULL | NULL | 3 | 33.33 | Using where |

**実行フロー**
1. シーケンシャルアクセスでparentテーブルにアクセスし、1行毎にname="Jhon"かをチェックする

## indexを貼っている場合
**実行計画**
| id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 1 | SIMPLE | parent | NULL | ref | idx_name | idx_name | 203 | const	| 1 | 100.00 | Using index |

**実行フロー**
1. ランダムアクセスでindexからname="Jhon"のアドレス番号を取得する
2. ランダムアクセスでparentテーブルにアクセスし、アドレス番号を使って直接取得する

## 常にindexのほうが早いのか？
アクセスするレコードがテーブル全体のレコード数の15%以下であれば、indexでランダムアクセスする方が早いと言われている
https://codezine.jp/article/detail/11425

**シーケンシャルアクセス**
- 表のサイズが大きいと、ディスクを回し切るまでの時間でかかり遅くなる
  - ex. 転送速度=20MB/秒としたとき、表サイズを1GBとすると、シーケンシャルアクセスで全部読み込むだけで50秒

**ランダムアクセス**
- シーケンシャルアクセスと比べてシークと回転待ちの時間で伝送速度が遅くなる
  - ex. ブロックサイズを8KB、1秒間にシークできる回数を100回とした場合、転送速度は800KB/秒(ディスクが指定した位置まで回っててくるまでの待ち時間もかかる)。

## indexのデータ数が増えると遅くならないか？
indexは木構造になっており、各データが他のデータへの参照を持っているため、データ量が増えたとしても特定のデータを検索する回数が少なくて済む。

### 同じ木構造の2分探索木で考える
**例**
- 身長のデータをindexにしたケースで考える

**2分探索木のルール**
- データを追加するとき、小さい値は左、大きい値は右につなげる
- 親がぶら下がっている子の参照(メモリ番号)を持つ

**何がいいのか？**
- 1/2ずつデータが絞られていくため、1024件のデータであっても10回の探索で済む(全体の1%程度)。
- 書き込みと読み出しどちらも高速にできる

```mermaid
graph TD;
    A[167]
    A --> B[165]
    A --> C[170]
    B --> D[163]
    B --> E[166]
    C --> F[168]
    C --> G[180]
```

## 結論
アクセスするレコードがテーブル全体のレコード数の15%以下であれば、indexの方が早い。indexを十分に使うには検索データを15%ぐらいになるまで絞り込むようにしよう。

**Comment by kakeru-one** [comment]
> もし結合キーにインデックスが貼られていれば、children テーブルを見に行かず、index を見に行くだけでいい

これは、InnoDBではクラスターインデックスなので、あんまり気にしなくて良さそうです。

**Reply by kakeru-one** [reply]
いや、関係ないか
